### PR TITLE
Lazy methods: mark JVM-generated exceptions as always available (fixes TG-774)

### PR DESCRIPTION
**Pull request by antlechner**
java.lang.NullPointerException et al are producable from Java ops
other than throw, and therefore weren't previously noticed. This adds
the exceptions that can be introduced by java_bytecode/java_bytecode_instrument.cpp
into the lazy methods loader's list of always-available classes.

Original PR:
https://github.com/diffblue/cbmc/pull/1467
(closed as it was based on a version of cbmc that broke test-gen)

### CODE REVIEW

**Comment by antlechner** [comment]
@smowton @thk123 @LAJW

**Comment by antlechner** [comment]
@thk123 About your comment on the previous PR: the changes are not visible in regression tests at the moment as CI uses platform options, which don't include `--java-throw-runtime-exceptions` yet. So any regression tests that check for this would have to be disabled until the parameter is enabled on platform. Do you think it's worth doing that?

**Comment by thk123** [comment]
@antlechner ~I meant the tests that already exist in cbmc-java that already have the--java-throw-runtime-exceptions flag on (for example `regression/cbmc-java/ClassCastException1`) if you add the lazy-methods param to them, I suppose they will fail~ except they won't have the models library! 

I guess the answer is to duplicate these tests (from `cbmc-java/ClassCastException`, `cbmc-java/NullPointerException1`... etc) and make them using `--lazy-methods` and models lib.

Not sure exactly what you mean re; CI uses platform options, doesn't it use whatever the test tells it to?

**Comment by antlechner** [comment]
@thk123 Sounds like the regession tests in cbmc work differently from the tests that CI uses for models-library. I haven't looked at the cbmc regression tests at all so far, so it will probably take me a few days to make those changes. As the changes aren't urgent as long as the parameter isn't enabled on platform, you can mark this as "changes requested" or "don't merge" for now, and I'll let you know when I've learnt about cbmc regression tests and added the ones that you suggested.

File: src/java_bytecode/ci_lazy_methods.cpp [hunk 1]
```diff
@@ -280,6 +280,13 @@ void ci_lazy_methodst::initialize_needed_classes(
   lazy_methods.add_needed_class("java::java.lang.String");
   lazy_methods.add_needed_class("java::java.lang.Class");
   lazy_methods.add_needed_class("java::java.lang.Object");
+  // Add the classes that can be produced by `java_bytecode_instrument`:
```
Comment (thk123) [inline src/java_bytecode/ci_lazy_methods.cpp]: Consider adding a comment to `java_bytecode_instrumentt::instrument_code` saying new runtime exceptions must be added to `lazy_methods` list of needed classes.

**Review by thk123 — CHANGES_REQUESTED** [review]
Please extend the runtime exception tests (found in `cbmc/regression/cbmc-java/`) to use lazy methods and the models libraryot exhibit this problem

**Comment by smowton** [comment]
Agree with @thk123 but don't worry, the time taken to add parameters like
this should be short given a quick tour of the test setup

**Comment by antlechner** [comment]
@tautschnig I just saw that you assigned this PR to me. It got delayed by company deadlines but I will revive it in a few days. I haven't forgotten about it.

**Comment by tautschnig** [comment]
Just went over all open PRs to figure out presumed ownership. Not intending to rush anyone, just tried to figure out where I might have to get my own act together...

**Comment by smowton** [comment]
Re-review-request may have been accidental (consequence of a push perhaps?) Just noting this still needs test cases.

**Comment by antlechner** [comment]
@smowton Yes, I just pushed the changes, didn't know that sent out review requests again, sorry. I'm working on adding tests now and trying to figure out how to make cbmc use (a limited version of) the models-library.

File: src/java_bytecode/java_class_loader.h [hunk 2]
```diff
@@ -96,6 +101,7 @@ class java_class_loadert:public messaget
 private:
   std::map<std::string, jar_filet> m_archives;
   std::vector<irep_idt> java_load_classes;
+  bool throw_runtime_exceptions;
```
Comment (mgudemann) [inline src/java_bytecode/java_class_loader.h]: I am not a big fan of uninitialized Boolean variables, having been bitten by difficult to track-down Heisenbugs. Would you mind giving a default value for this and the above?

**Review by mgudemann — CHANGES_REQUESTED** [review]

File: src/java_bytecode/ci_lazy_methods.cpp (hunk 1 above)
Comment (antlechner) [reply src/java_bytecode/ci_lazy_methods.cpp]: I added a comment to the `instrument_code` function. Could you check if it makes sense and conforms to Doxygen standards?

File: src/java_bytecode/java_class_loader.h (hunk 2 above)
Comment (antlechner) [reply src/java_bytecode/java_class_loader.h]: I gave both of them `false` as a default value. (Seemed like the safer option as the test-gen `--java-throw-runtime-exceptions` parameter isn't/can't be enabled on platform yet.)

**Comment by antlechner** [comment]
@smowton I added test cases for all five exceptions. I followed your advice of using a very small subset of the models-library in the tests: basically just empty classes that follow the class hierarchy, e.g. a `RuntimeException` class that extends `Exception`, and an `ArithmeticException` class that extends `RuntimeException`.

@thk123 Is this consistent with your idea of using the models-library in tests? (Sorry it's been a while since I opened this PR, I was too busy working on generics in the past weeks.)

I ran all five tests locally before and after the changes. All failed before, and passed after.

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 3]
```diff
@@ -354,7 +354,17 @@ void java_bytecode_instrumentt::prepend_instrumentation(
...
+/// The list of runtime exceptions that can be thrown without having been
+/// declared explicitly using 'throw new' is:
+/// ArithmeticException
+/// ArrayIndexOutOfBoundsException
+/// ClassCastException
+/// NegativeArraySizeException
+/// NullPointerException
+/// If this list grows in the future, any new classes will have to both be
+/// handled by this function and added to the list of classes that should
+/// always be loaded (by the Java class loader and by lazy methods).
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Looks good - but could you put explicit source locations (e.g. `class_name::function`) so people know where to look

**Review by thk123 — COMMENTED** [review]
The tests look good - one thing I might suggest is would it be possible to have one list `exception_classes` so it only has to be updated in one place. Or perhaps this should be done with @mgudemann recent PR with specifying the ArrayList etc.

**Review by smowton — APPROVED** [review]
Looks ideal, thanks for adding the tests!

**Comment by antlechner** [comment]
@thk123 Good point, it turned out I could just use @mgudemann 's new `java_load_classes` variable instead of giving two classes a `bool java_throw_runtime_exceptions` variable. This way I only have the list of necessary classes defined in one place now. I chose `java_bytecode_instrument`, as this is where the exceptions are handled, and I thought that way the exception-related logic would all be in one place. However, the `java_bytecode_instrumentt` class is not publicly accessible, so I had to define the new variable using the `extern` keyword.

@mgudemann The lines that you commented on disappeared (along with pretty much all other initial changes). I changed a `for` loop that you wrote recently to a `std::vector.insert` as @smowton suggested it would be nice to avoid using two loops there (the one that was there plus the one I added).

In summary, tests haven't changed but please re-review the actual code changes from scratch. :)

File: src/java_bytecode/java_bytecode_language.cpp [hunk 4]
```diff
@@ -29,8 +29,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "java_bytecode_parser.h"
 #include "java_class_loader.h"
 #include "java_utils.h"
-#include <java_bytecode/ci_lazy_methods.h>
-#include <java_bytecode/generate_java_generic_type.h>
+#include "ci_lazy_methods.h"
+#include "java_bytecode/generate_java_generic_type.h"
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_language.cpp]: What is the purpose of this change? If intentional it should be consistent - since both files are in `java_bytecode` they either should both prefix it (which I think is what the coding guidelines say) or neither (which seems more consistent with the other includes in this file)

**Review by thk123 — APPROVED** [review]

File: src/java_bytecode/java_bytecode_language.cpp (hunk 4 above)
Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp]: I'd drop this as unrelated cleanup

Comment (antlechner) [reply src/java_bytecode/java_bytecode_language.cpp]: I got rid of this (reset it to what it was), sorry for the confusion. I originally added more includes here, tried to change the existing ones to make them more consistent, then ended up in a half-finished state and forgot to change them back.

**Review by mgudemann — APPROVED** [review]
LGTM

**Review by mgudemann — APPROVED** [review]
LGTM
